rust/worker/src/execution/operators/brute_force_knn.rs

```diff
@@ -17,45 +21,103 @@ pub struct BruteForceKnnOperator {}
 /// * `distance_metric` - The distance metric to use.
 #[derive(Debug)]
 pub struct BruteForceKnnOperatorInput {
-    pub data: Vec<Vec<f32>>,
+    pub data: DataChunk,
     pub query: Vec<f32>,
     pub k: usize,
     pub distance_metric: DistanceFunction,
 }
 
 /// The output of the brute force k-nearest neighbors operator.
 /// # Parameters
+/// * `data` - The vectors to query against. This is a mask against the `data` input.
 /// * `indices` - The indices of the nearest neighbors. This is a mask against the `query_vecs` input.
 /// One row for each query vector.
 /// * `distances` - The distances of the nearest neighbors.
 /// One row for each query vector.
 #[derive(Debug)]
 pub struct BruteForceKnnOperatorOutput {
+    pub data: DataChunk,
     pub indices: Vec<usize>,
     pub distances: Vec<f32>,
 }
 
 pub type BruteForceKnnOperatorResult = Result<BruteForceKnnOperatorOutput, ()>;
 
+#[derive(Debug)]
+struct Entry {
+    index: usize,
+    distance: f32,
+}
+
+impl Ord for Entry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.distance == other.distance {
+            Ordering::Equal
+        } else if self.distance > other.distance {
+            // This is a min heap, so we need to reverse the ordering.
+            Ordering::Less
+        } else {
+            // This is a min heap, so we need to reverse the ordering.
+            Ordering::Greater
+        }
+    }
+}
+
+impl PartialOrd for Entry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for Entry {
+    fn eq(&self, other: &Self) -> bool {
+        self.distance == other.distance
+    }
+}
+
+impl Eq for Entry {}
+
 #[async_trait]
 impl Operator<BruteForceKnnOperatorInput, BruteForceKnnOperatorOutput> for BruteForceKnnOperator {
     type Error = ();
 
     async fn run(&self, input: &BruteForceKnnOperatorInput) -> BruteForceKnnOperatorResult {
-        // We could use a heap approach here, but for now we just sort the distances and take the
-        // first k.
-        let mut sorted_indices_distances = input
-            .data
-            .iter()
-            .map(|data| input.distance_metric.distance(&input.query, data))
-            .enumerate()
-            .collect::<Vec<(usize, f32)>>();
-        sorted_indices_distances.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap());
-        let (sorted_indices, sorted_distances) = sorted_indices_distances
-            .drain(..cmp::min(input.k, input.data.len()))
-            .unzip();
+        let mut heap = BinaryHeap::with_capacity(input.k);
+        let data_chunk = &input.data;
+        for data in data_chunk.iter() {
+            let embedding_record = data.0;
+            let index = data.1;
 
+            let embedding = match &embedding_record.embedding {
```

Comment:
This is ugly, and part of the reason I'm a bit hesitant to pull the trigger on DataChunk. Why does this operator need to be aware of the fact that the embedding could be empty here? It can only be empty if its a delete. I think we should rethink this.